### ISSUE TITLE
Work needed to enable `useOtelForInternalMetrics` by default

### ISSUE THREAD

**Issue by codeboten** [open-telemetry/opentelemetry-collector#7454]
This issue is to discuss the work needed to enable using the OTel SDK by default. 

Related issues:

- [ ] https://github.com/open-telemetry/opentelemetry-collector/issues/816

**Comment by codeboten** [comment]
@jmacd I believe this PR will help w/ #3: https://github.com/open-telemetry/opentelemetry-collector/pull/7483

The issue is that currently both OC and OT are using prometheus as a bridge to allow components to continue using  opencensus instrumentation. The change I'm proposing disables the OC configuration when `useOtelForInternalMetrics` is enabled and relies on the oc to otel bridge instead.

**Comment by mx-psi** [comment]
We may need to address #7517 before closing this

**Comment by mx-psi** [comment]
The `check_collector_pipeline` functionality in the health check extension depends on OpenCensus, see [here]( https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/e42f386e07d6b2099ff82d664dc7e54d6f2af0b4/extension/healthcheckextension/exporter.go#L14). We would need to either deprecate this functionality or update it to use OTel before enabling `useOtelForInternalMetrics`

**Comment by codeboten** [comment]
With the PR for `_total` suffix merging yesterday, I started looking at the second issue @jmacd mentioned in his comment above. With opencensus telemetry, service labels are applied on each metric:

```
otelcol_exporter_queue_capacity{exporter="otlp",service_instance_id="aae63e2f-cad2-4224-bbb3-806689b0a2d0",service_name="otelcontribcol",service_version="0.89.0-dev"} 1000
```

With `otelForInternalMetrics` enabled, all that information is set in the target_info:

```
otelcol_target_info{service_instance_id="73a547db-f135-45c4-870e-cde704b3cf7d",service_name="otelcontribcol",service_version="0.89.0-dev"} 1
```

I'm not sure that there's a good mechanism today for applying info from target_info as resource attributes. @dashpole pointed me to a spec change proposed to address this for prometheus exporters https://github.com/open-telemetry/opentelemetry-specification/pull/3761 (thanks David!)

I don't know what the right way to move forward with this change is with the current state. I did test the OTLP exporter for internal metrics and the resource attributes contained the service information I expected. The following is the output from the Go stdout exporter:

```
  "Resource": [
    {
      "Key": "service.instance.id",
      "Value": {
        "Type": "STRING",
        "Value": "cdc7786f-fcfc-4530-98cd-7278e6dbcf5c"
      }
    },
    {
      "Key": "service.name",
      "Value": {
        "Type": "STRING",
        "Value": "otelcontribcol"
      }
    },
    {
      "Key": "service.version",
      "Value": {
        "Type": "STRING",
        "Value": "0.89.0-dev"
      }
    }
  ],
```

In other words, for end users that want OTLP for their collector metrics (like @jmacd's original scenario), this would be a good replacement for what is only available today w/ prometheus metrics (with OC). For users that want to continue using the Prometheus export functionality, I'm not sure what to recommend.

**Comment by dashpole** [comment]
Do you know why `otelcol_target_info` has the otelcol prefix?  We would really like to keep the name unmodified, if possible.

**Comment by codeboten** [comment]
> Do you know why `otelcol_target_info` has the otelcol prefix? We would really like to keep the name unmodified, if possible.

It's prefixed here https://github.com/open-telemetry/opentelemetry-collector/blob/445960b82fb4f4a979bfa5e6ce88aa754d22b92d/service/internal/proctelemetry/config.go#L206C1-L206C84

**Comment by dashpole** [comment]
I think we can wrap [Collect](https://github.com/prometheus/client_golang/blob/80d3f0b5b36c891016370423516e353b1a5f0a55/prometheus/collector.go#L62) to add resource attributes as additional labels.  That is what the prometheus exporter implements.  In Collect, we would need to wrap each Metric, and override the Write function.  There, we would need to add our additional labels.

It would end up looking similar to https://github.com/prometheus/client_golang/blob/80d3f0b5b36c891016370423516e353b1a5f0a55/prometheus/metric.go#L172, which wraps another metric and adds exemplars to it.

**Comment by jpkrohling** [comment]
During the SIG call, I mentioned that Java implemented a double-emit so that an application could use both a new and an old semantic convention at the same time, to make it easier to migrate between versions without breaking other parts of the observability pipeline. That feature is described here: https://github.com/open-telemetry/semantic-conventions/blob/main/docs/http/README.md

**Comment by codeboten** [comment]
@jpkrohling I guess this would raise the question of what's the correct thing to emit since there are a couple of differences between the current metrics and what's are the correct metrics:

1. `_total` suffix should be emitted
2. `target_info` should be present and new metrics should not include resource labels

Are you proposing that we have two endpoints that emit these metrics at the same time? I'm not sure what the deprecation story would be here, since it would be hard for users to know this change is coming. We saw changing metrics surprise users when we tried to change the default behaviour of counter metrics to include the `_total` suffix in the past.

Maybe the solution is to have a configuration option to use the legacy format?

**Comment by codeboten** [comment]
@dashpole I followed up on your question about the prefix, and made the change: https://github.com/open-telemetry/opentelemetry-collector/pull/8988

I wanted to answer another question from the SIG meeting which was whether the prometheus receiver was able to use this `target_info`. Before the change, it would treat `otelcol_target_info` as any other metric. After #8988, the prometheus receiver to use `target_info` and add its contents in the emitted telemetry. The following is the resource attributes output of a prometheus receiver scraping the otel collector internal metrics and emitting them as a debug exporter:

```
2023-11-23T12:04:37.361-0800	info	ResourceMetrics #0
Resource SchemaURL:
Resource attributes:
     -> service.name: Str(self)
     -> service.instance.id: Str(127.0.0.1:8889)
     -> net.host.port: Str(8889)
     -> http.scheme: Str(http)
     -> service_instance_id: Str(59dee129-67c4-4981-b077-e4e04aa3e4e6)
     -> service_name: Str(otelcorecol)
     -> service_version: Str(0.89.0-dev)
```

One more question that ^^^ raises is whether or not this is correct. As you can see there is both a `service.name` and a `service_name` attribute which are different, service.name is the prometheus job name configured in the receiver, service_name is the collector's target_info details.

**Comment by jpkrohling** [comment]
> `target_info` should be present and new metrics should not include resource labels
> ...
> Maybe the solution is to have a configuration option to use the legacy format?

Perhaps a feature flag can control the three possible scenarios? No target info + all resource attributes on metrics, target info and minimal set of resource attributes on metrics, target info and all resource attributes.

The third scenario would help people migrate from the previous to the new one, as they can just ignore the extra resource attributes when building the integration. Once they are ready, they can move to the scenario 2, which is what will be supported for the long term.

**Comment by codeboten** [comment]
> > `target_info` should be present and new metrics should not include resource labels
> > ...
> > Maybe the solution is to have a configuration option to use the legacy format?
> 
> Perhaps a feature flag can control the three possible scenarios? No target info + all resource attributes on metrics, target info and minimal set of resource attributes on metrics, target info and all resource attributes.
> 
> The third scenario would help people migrate from the previous to the new one, as they can just ignore the extra resource attributes when building the integration. Once they are ready, they can move to the scenario 2, which is what will be supported for the long term.

Are you suggesting a single gate that could be used for all 3 scenarios? If so then I would recommend we:
1. wait until we can produce metrics that are identical using OTel,
2. enable otel by default and deprecate this feature gate
3. create a `new` gate that allows users to change the output (open to ideas for name here)
4. enable this `new` gate as the default some time in the future

In other words, I would like to decouple people having to change the metrics they're using from enabling otel. That being said, it's a bit weird to enable otel with the old style metrics, but at least it unblocks that work

**Comment by dashpole** [comment]
I would guess a week or two?  I don't think it will be hard to implement.
